File: app/ViRulerView.m
/*
 * Copyright (c) 2008-2012 Martin Hedenfalk <martin@vicoapp.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *

#import "ViRulerView.h"
#include "ViFold.h"
#import "ViTextView.h"
#import "ViThemeStore.h"
//#import "NSObect+SPInvocationGrabbing.h"
#include "logging.h"

#define DEFAULT_THICKNESS   22.0
#define RULER_MARGIN        5.0

@implementation ViRulerView

- (id)initWithScrollView:(NSScrollView *)aScrollView
{
	if ((self = [super initWithScrollView:aScrollView orientation:NSVerticalRuler]) != nil) {
		_backgroundColor = [NSColor colorWithDeviceRed:(float)0xED/0xFF
							 green:(float)0xED/0xFF
							  blue:(float)0xED/0xFF
							 alpha:1.0];

		[self setClientView:[[self scrollView] documentView]];
	}

	return self;
}

- (void)dealloc
{
	[[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)setRelativeLineNumbers:(BOOL)flag
{
	if (_lineNumberView)
		[_lineNumberView setRelative:flag];
}

- (void)lineViewBoundsDidChange:(NSNotification *)aNotification
{
	[self setRuleThickness:[_lineNumberView requiredThickness]];
}

- (void)resetTextAttributes
{
	[_lineNumberView resetTextAttributes];
}

#pragma mark -
#pragma mark NSRulerView interface

- (void)setClientView:(NSView *)aView
{
	[super setClientView:aView];

	if (aView != nil && [aView isKindOfClass:[ViTextView class]]) {
		if (_lineNumberView) {
			[_lineNumberView setTextView:(ViTextView *)aView];
		} else {
			_lineNumberView = [[ViLineNumberView alloc] initWithTextView:(ViTextView *)aView
														 backgroundColor:_backgroundColor];

			[[NSNotificationCenter defaultCenter] addObserver:self
													 selector:@selector(subviewFrameDidChange:)
														 name:NSViewFrameDidChangeNotification
													   object:_lineNumberView];

			[self addSubview:_lineNumberView];
			[_lineNumberView updateViewFrame];
		}

		if (_foldMarginView) {
			[_foldMarginView setTextView:(ViTextView *)aView];
		} else {
			_foldMarginView = [[ViFoldMarginView alloc] initWithTextView:(ViTextView *)aView];

			[[NSNotificationCenter defaultCenter] addObserver:self
													 selector:@selector(subviewFrameDidChange:)
														 name:NSViewFrameDidChangeNotification
													   object:_foldMarginView];

			[self addSubview:_foldMarginView];
			[_foldMarginView updateViewFrame];
		}

		[self updateRuleThickness];
	}
}

- (void)subviewFrameDidChange:(NSNotification *)aNotification
{
	[self updateRuleThickness];
}

- (void)updateRuleThickness
{
	CGFloat newThickness = _lineNumberView.frame.size.width + _foldMarginView.frame.size.width;
	
	if (newThickness != [self ruleThickness]) {
		[[self nextRunloop] setRuleThickness:newThickness];
	}
}

- (void)drawHashMarksAndLabelsInRect:(NSRect)aRect
{
	NSRect bounds = [self bounds];
	NSRect visibleRect = [[[self scrollView] contentView] bounds];

	[_backgroundColor set];
	NSRectFill(bounds);

	[[NSColor colorWithCalibratedWhite:0.58 alpha:1.0] set];
	[NSBezierPath strokeLineFromPoint:NSMakePoint(NSMaxX(bounds) - 0.5, NSMinY(bounds))
				  toPoint:NSMakePoint(NSMaxX(bounds) - 0.5, NSMaxY(bounds))];

	[_lineNumberView drawLineNumbersInRect:aRect visibleRect:visibleRect];
	[_foldMarginView drawFoldsInRect:aRect visibleRect:visibleRect];
}

#pragma mark -
#pragma mark Mouse handling

- (void)mouseUp:(NSEvent *)theEvent
{
	NSPoint upPoint = [self convertPoint:theEvent.locationInWindow fromView:nil];
	if (upPoint.x > _lineNumberView.frame.size.width) {
		[_foldMarginView foldMarginMouseUp:theEvent];
	}
}

- (void)mouseDown:(NSEvent *)theEvent
{
	NSPoint downPoint = [self convertPoint:theEvent.locationInWindow fromView:nil];
	if (downPoint.x <= _lineNumberView.frame.size.width) {
		[_lineNumberView lineNumberMouseDown:theEvent];
	}
}

- (void)mouseDragged:(NSEvent *)theEvent
{
	NSPoint dragPoint = [self convertPoint:theEvent.locationInWindow fromView:nil];
	if (dragPoint.x <= _lineNumberView.frame.size.width) {
		[_lineNumberView lineNumberMouseDragged:theEvent];
	}
}

@end
 */

File: app/Ruler/ViRulerView.swift
//
//  ViRulerView.swift
//  vico
//
//  Created by Antonio Salazar Cardozo on 6/28/15.
//
//

import Foundation

class ViRulerView: NSRulerView {
    static let rerenderingDocumentNotifications = [ViFoldsChangedNotification, ViFoldOpenedNotification, ViFoldClosedNotification]
    
    private let backgroundColor =
        NSColor(
            deviceRed: (0xED / 0xFF) as CGFloat,
            green: (0xED / 0xFF) as CGFloat,
            blue: (0xED / 0xFF) as CGFloat,
            alpha: 1.0
        )
    private let marginColor =
        NSColor(calibratedWhite: 0.58, alpha: 1.0)
    
    private var textView: ViTextView? {
        willSet {
            if let textView = self.textView {
                NSNotificationCenter.defaultCenter().removeObserver(self,
                    name: ViTextStorageChangedLinesNotification,
                    object: textView.textStorage)
                
                NSNotificationCenter.defaultCenter().removeObserver(self,
                    name: ViCaretChangedNotification,
                    object: textView)
                
                if newValue?.document != textView.document {
                    for notification in ViRulerView.rerenderingDocumentNotifications {
                        NSNotificationCenter.defaultCenter().removeObserver(self,
                            name: notification,
                            object: textView.document)
                    }
                }
            }
        }
        didSet {
            guard let textView = self.textView else {
                fatalError("Tried to use ViRulerView without a text view.")
            }
            
            for var helper in helpers {
                helper.textView = textView
            }
            
            NSNotificationCenter.defaultCenter().addObserver(self,
                selector: "textStorageDidChangeLines:",
                name: ViTextStorageChangedLinesNotification,
                object: textView.textStorage)
            
            NSNotificationCenter.defaultCenter().addObserver(self,
                selector: "ruleThicknessUpdateNeeded:",
                name: ViCaretChangedNotification,
                object: textView)
            
            if oldValue?.document != textView.document {
                for notification in ViLineNumberView.rerenderingDocumentNotifications {
                    NSNotificationCenter.defaultCenter().addObserver(self,
                        selector: "ruleThicknessUpdateNeeded:",
                        name: notification,
                        object: textView.document)
                }
            }
        }
    }
    
    private var _helpers: [ViRulerHelper] = []
    private var helpers: [ViRulerHelper] {
        get {
            if let textView = self.textView where _helpers.isEmpty {
                _helpers = helpersForTextView(textView, backgroundColor: backgroundColor)
                
                for helper in _helpers {
                    let view = helper.helperView
                    
                    self.addSubview(view)
                    NSNotificationCenter.defaultCenter().addObserver(self,
                        selector: "ruleThicknessUpdateNeeded:",
                        name: NSViewFrameDidChangeNotification,
                        object: view
                    )
                }
            }
            
            return _helpers
        }
    }
    
    override var clientView: NSView? {
        didSet {
            textView = clientView as? ViTextView
        }
    }
    override var scrollView: NSScrollView? {
        didSet {
            if let scrollView = self.scrollView, documentView = scrollView.documentView as? NSView {
                clientView = documentView
            }
        }
    }
    
    init(scrollView: NSScrollView) {
        super.init(scrollView: scrollView, orientation: .VerticalRuler)
    }

    required init?(coder: NSCoder) {
        fatalError("ViRulerView currently doesn't support coding/decoding.")
    }
    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    @objc func ruleThicknessUpdateNeeded(notification: NSNotification) {
        updateRuleThickness()
    }
    
    @objc func textStorageDidChangeLines(notification: NSNotification) {
        needsDisplay = true
    }
    
    private func updateRuleThickness() {
        let newThickness = helpers.reduce(0, combine: { $0 + $1.helperView.frame.width })
        
        if newThickness != ruleThickness {
            NSOperationQueue.mainQueue().addOperationWithBlock({ [weak self] () -> Void in
                self?.ruleThickness = newThickness
            })
        }
        
        needsDisplay = true
    }
    
    private func fillBackground() {
        backgroundColor.set()
        NSRectFill(bounds)
    }
    
    private func drawMargin() {
        marginColor.set()
        
        let marginThickness: CGFloat = 0.5
        NSBezierPath.strokeLineFromPoint(
            NSMakePoint(NSMaxX(bounds) - marginThickness, NSMinY(bounds)),
            toPoint: NSMakePoint(NSMaxX(bounds) - marginThickness, NSMaxY(bounds))
        )
    }
    
    override func drawHashMarksAndLabelsInRect(rect: NSRect) {
        if let visibleRect = scrollView?.contentView.bounds {
            fillBackground()
            drawMargin()
            
            for helper in helpers {
                helper.drawInRect(rect, visibleRect: visibleRect)
            }
        }
    }
    
    func resetTextAttributes() {
        for helper in helpers {
            helper.resetTextAttributes()
        }
    }
}